### ISSUE TITLE
Random crashes in the log with OTP 21 and 22

### ISSUE THREAD

**Issue by NobbZ** [webmachine/webmachine#301]
I'm trying to get into `webmachine` and see random crashes all over the place, though the "Hello, World!" example seems to serve the expected page while crashing in the "background"…

Crashes happen with requests happening or without and all of them that I have observed so far do refer to a `baf argument` when calling `apply/3`.

Those crashes do not happen on OTP 20.

Versions of `webmachine` and `mochiweb` per `rebar.lock`:

```erl
[{<<"mochiweb">>,{pkg,<<"mochiweb">>,<<"2.18.0">>},1},
 {<<"webmachine">>,{pkg,<<"webmachine">>,<<"1.11.1">>},0}]
```

A randomly picked crash report:

```plain
=CRASH REPORT==== 26-Jan-2020::09:16:25.880127 ===
  crasher:
    initial call: mochiweb_acceptor:init/4
    pid: <0.7921.7>
    registered_name: []
    exception error: bad argument
      in function  apply/3
         called as apply({mochiweb_request,
                             [#Port<0.74>,
                              [{recbuf,8192}],
                              'POST',"/cgi-bin/ViewLog.asp",
                              {1,1},
                              {7,
                               {"host",
                                {'Host',"104.168.200.193:80"},
                                {"connection",
                                 {'Connection',"keep-alive"},
                                 {"accept-encoding",
                                  {'Accept-Encoding',"gzip, deflate"},
                                  {"accept",{'Accept',"*/*"},nil,nil},
                                  nil},
                                 {"content-length",
                                  {'Content-Length',"227"},
                                  nil,
                                  {"content-type",
                                   {'Content-Type',
                                       "application/x-www-form-urlencoded"},
                                   nil,nil}}},
                                {"user-agent",
                                 {'User-Agent',"python-requests/2.20.0"},
                                 nil,nil}}}]},
                         get,
                         [socket])
      in call from mochiweb_http:after_response/2 (/home/nmelzer/projects/erlang/nobbz_dev/_build/default/lib/mochiweb/src/mochiweb_http.erl, line 173)
    ancestors: [webmachine_mochiweb,nobbz_dev_sup,<0.588.0>]
    message_queue_len: 0
    messages: []
    links: [<0.592.0>,#Port<0.74>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 4247
  neighbours:
```

**Comment by beerriot** [comment]
I think it is probably caused by the [removal of parameterized modules in OTP 21](https://www.erlang.org/downloads/21.0)

> "Tuple calls" have been removed from the run-time system.

It should be fixable by bumping the mochiweb dependency to either v2.19.0, which added the `-compile(tuple_calls).` directive where necessary in mochiweb, or v2.20.0, which removed tuple calls.

The follow-up question will be whether we finally removed all parameterized modules from webmachine itself.

**Comment by NobbZ** [comment]
Oh, sorry, I have not paid attention to the source of the report.

When I'm back home, I'll try to do a better job. Perhaps I can see some crashes which are more obviously related to `webmachine`.

Perhaps I'll even try ovverriding the `mochiweb` dependency.

**Comment by samwar** [comment]
@beerriot This is most definitely the issue. The current tagged version of webmachine is using mochiweb 2.18.0, which coincidentally is the only version published in hex. I've commented on https://github.com/mochi/mochiweb/issues/200 so hopefully that will get resolved soon. In the meantime the dependency for webmachine could be changed to github and a new version could be tagged and released. That would make it possible to cleanly upgrade to erlang 21.

**Comment by samwar** [comment]
I've made a [PR](https://github.com/webmachine/webmachine/pull/302) to use the latest git version of mochiweb. Let's hope either you @beerriot or @seancribbs merges it and then cuts a new release of webmachine.

**Comment by NobbZ** [comment]
You can't release to hex when you have a git dependency though… Please be aware of that. And to be honest, I do prefer to use dependencies from hex rather than version control…

Testing things is a different thing, but for the "product" properly released libraries are a surplus.

**Comment by samwar** [comment]
I'll be happy to get a new PR out there if a new version of mochiweb is published. I don't disagree with you on having properly released libraries somewhere. At the very least, the the mochiweb version is pinned to a commit, and those are immutable.

**Comment by beerriot** [comment]
Just cleaning up a bit. It seems like this should have closed when #302 merged, yes?
